packages/db-migrations/databases/fxa/patches/patch-133-134.sql

```diff
@@ -0,0 +1,138 @@
+      SET signInCodesDeleted = ROW_COUNT();
+    END IF;
+
+    IF maxTokenAge > 0 THEN
+      DELETE FROM accountResetTokens   WHERE createdAt < curTime - maxTokenAge ORDER BY createdAt LIMIT 10000;
+      SET accountResetTokensDeleted = ROW_COUNT();
+      DELETE FROM passwordForgotTokens WHERE createdAt < curTime - maxTokenAge ORDER BY createdAt LIMIT 10000;
+      SET passwordForgotTokensDeleted = ROW_COUNT();
+      DELETE FROM passwordChangeTokens WHERE createdAt < curTime - maxTokenAge ORDER BY createdAt LIMIT 10000;
+      SET passwordChangeTokensDeleted = ROW_COUNT();
+
+      -- Pruning session tokens is complicated because:
+      --   * we can't prune them if there is an associated device record, and
+      --   * we have to delete from both sessionTokens and unverifiedTokens tables, and
+      --   * MySQL won't allow `LIMIT` to be used in a multi-table delete.
+      -- To achieve all this in an efficient manner, we prune tokens within a specific
+      -- time window rather than using a `LIMIT` clause.  At the end of each run we
+      -- record the new lower-bound on creation time for tokens that might have expired.
+      START TRANSACTION;
+
+      -- Step 1: Find out how far we got on previous iterations.
+      SELECT @pruneFrom := value FROM dbMetadata WHERE name = 'prunedUntil';
+
+      -- Step 2: Calculate what timestamp we will reach on this iteration
+      -- if we purge a sensibly-sized batch of tokens.
+      -- N.B. We deliberately do not filter on whether the token has
+      -- a device here.  We want to limit the number of tokens that we
+      -- *examine*, regardless of whether it actually delete them.
+      SELECT @pruneUntil := MAX(createdAt) FROM (
+        SELECT createdAt FROM sessionTokens
+        WHERE createdAt >= @pruneFrom AND createdAt < curTime - maxTokenAge
+        ORDER BY createdAt
+        LIMIT 10000
+      ) AS candidatesForPruning;
+
+      -- This will be NULL if there are no expired tokens,
+      -- in which case we have nothing to do.
+      IF @pruneUntil IS NOT NULL THEN
+
+        -- Step 3: relay impacted accounts back to token pruner
+        SELECT distinct st.uid
+        from sessionTokens AS st
+        LEFT JOIN unverifiedTokens AS ut
+        ON st.tokenId = ut.tokenId
+        WHERE st.createdAt > @pruneFrom
+        AND st.createdAt <= @pruneUntil
+        AND NOT EXISTS (
+            SELECT sessionTokenId FROM devices
+            WHERE uid = st.uid AND sessionTokenId = st.tokenId
+        );
+
+        -- Step 4: Prune sessionTokens and unverifiedTokens tables.
+        -- Here we *do* filter on whether a device record exists.
+        -- We might not actually delete any tokens, but we will definitely
+        -- be able to increase 'prunedUntil' for the next run.
+        DELETE st, ut
+        FROM sessionTokens AS st
+        LEFT JOIN unverifiedTokens AS ut
+        ON st.tokenId = ut.tokenId
+        WHERE st.createdAt > @pruneFrom
```

Comment:
Also `>=` here?

Reply:
I didn't write this logic initially, but I think it is fine as is. The `@prunedFrom` is always exclusive and `@prunedUntil` is inclusive. To me this seems okay.

Reply:
I just revisited this, and I now see that it doesn't seem to be inline with how prunedUtil is determined. I will bring these in alignment.